sig-azure/0018-20180711-azure-availability-zones.md

```diff
@@ -0,0 +1,279 @@
+
+### StorageClass
+
+Note that the above interfaces are only applied to AzureDisk persistent volumes, not StorageClass. For AzureDisk StorageClass, we should add a few new options for zone-aware and [topology-aware](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/volume-topology-scheduling.md) provisioning. The following three new options will be added in AzureDisk StorageClass:
+
+- `zoned`: indicates whether new disks are provisioned with AZ. Default is `true`.
+- `zone` and `zones`: indicates which zones should be used to provision new disks (zone-aware provisioning). Only can be set if `zoned` is not false and `allowedTopologies` is not set.
+- `allowedTopologies`: indicates which topologies are allowed for [topology-aware](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/volume-topology-scheduling.md) provisioning. Only can be set if `zoned` is not false and `zone`/`zones` are not set.
+
+An example of zone-aware provisioning storage class is:
+
+```yaml
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  annotations:
+  labels:
+    kubernetes.io/cluster-service: "true"
+  name: managed-premium
+parameters:
+  kind: Managed
+  storageaccounttype: Premium_LRS
+  # only one of zone and zones are allowed
+  zone: "centralus-1"
+  # zones: "centralus-1,centralus-2,centralus-3"
+provisioner: kubernetes.io/azure-disk
+```
+
+Another example of topology-aware provisioning storage class is:
+
+```yaml
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  annotations:
+  labels:
+    kubernetes.io/cluster-service: "true"
+  name: managed-premium
+parameters:
+  kind: Managed
+  storageaccounttype: Premium_LRS
+provisioner: kubernetes.io/azure-disk
+allowedTopologies:
+- matchLabelExpressions:
+  - key: failure-domain.beta.kubernetes.io/zone
+    values:
+    - centralus-1
+    - centralus-2
+```
+
+AzureDisk can only be created with one specific zone, so if multiple zones are specified in the storage class, then new disks will be provisioned with zone chosen arbitrarily among them.
+
+If no zones are specified and `zoned` is not false, then new disks will be provisioned with zone chosen by round-robin across all active zones, which means
+
+- If there are no zoned nodes, then an `no zoned nodes` error will be reported
+- Zoned AzureDisk will only be provisioned when there are zoned nodes
+- If there are multiple zones, then those zones are chosen by round-robin
+
+Note that there are risks if the cluster is running with both zoned and non-zoned nodes. In such case, zoned AzureDisk can't be attached to non-zoned nodes. This means
+
+- new pods with zoned AzureDisks are always scheduled to zoned nodes
+- old pods using non-zoned AzureDisks can't be scheduled to zoned nodes
+
+So if users are planning to migrate workloads to zoned nodes, old AzureDisks should be recreated (probably backup first and restore to the new one).
+
```

Comment:
Maybe you can add a section to call out the work needed in AzureDisks's `Provision()` to support Topology Aware Dynamic provisioning feature. This can cover the considerations in this [EBS PR](https://github.com/kubernetes/kubernetes/pull/65730) which will also be added for GCP.

Reply:
Thanks. I thought active zones means what you have said above. But for clarity, I also added this in the new commit.

For AzureDisks's Provision(), it should be the same logic as EBS. The EBS PR is indeed a good refer for azuredisk.